proposals/0061-fair-congestion-control/simulation/src/main.rs

```diff
@@ -0,0 +1,565 @@
+    clock: u64,
+    reset_counter: u64,
+    freq: u64,
+    is_active: bool,
+}
+
+impl LocalFeeMarket {
+    fn new(reset_counter: u64) -> Self {
+        Self {
+            required_fee_rate: MINIMUM_BASE_FEE_RATE,
+            reserved_fee: INITIAL_RESERVED_FEE,
+            clock: 0,
+            reset_counter,
+            freq: 0,
+            is_active: false,
+        }
+    }
+}
+
+#[derive(Default, Debug)]
+struct BaseFeeTracker<const POLICY: Policy> {
+    clock: u64,
+    reset_counter: u64,
+    is_congested: bool,
+    fee_markets: BTreeMap<Addr, LocalFeeMarket>,
+    active_txs: BTreeSet<TxId>,
+    nonconflicting_group_count: u64,
+    recent_addrs: VecDeque<Vec<Addr>>,
+    rewarded_cu: u64,
+    total_supplied_fee: u64,
+}
+
+const MINIMUM_BASE_FEE_RATE: u64 = 5000;
+const CU_TO_POWER: f64 = 50_000.0;
+const INITIAL_RESERVED_FEE: u64 = 0;
+
+#[derive(Debug, PartialEq)]
+enum MeasureError {
+    AlreadyMeasuring,
+    AlreadyActiveAddress,
+    NoAddress,
+    NotMeasured,
+    TooManyActiveThreadCount,
+    InsufficientSuppliedFee(u64, u64),
+}
+use MeasureError::*;
+
+#[derive(Debug, PartialEq, Eq)]
+struct Policy {
+    congestion_threshold: usize,
+    recent_tx_count: usize,
+    maximum_thread_count: usize,
+    exponential_rate: u64, // can't directly use f64 yet in consts...
+}
+
+impl Policy {
+    const fn new() -> Self {
+        Self {
+            congestion_threshold: 0,
+            recent_tx_count: 5,
+            maximum_thread_count: 5,
+            exponential_rate: 2_f64.to_bits(),
+        }
+    }
+
+    const fn congestion_threshold(mut self, congestion_threshold: usize) -> Self {
+        self.congestion_threshold = congestion_threshold;
+        self
+    }
+
+    const fn recent_tx_count(mut self, recent_tx_count: usize) -> Self {
+        self.recent_tx_count = recent_tx_count;
+        self
+    }
+
+    const fn maximum_thread_count(mut self, maximum_thread_count: usize) -> Self {
+        self.maximum_thread_count = maximum_thread_count;
+        self
+    }
+
+    fn exponential_rate(&self) -> f64 {
+        f64::from_bits(self.exponential_rate)
+    }
+}
+
+impl<const POLICY: Policy> BaseFeeTracker<POLICY> {
+    fn start_measuring(&mut self, tx: &Tx) -> Result<(), MeasureError> {
+        let updated_active_tx_count = self.active_txs.len() + 1;
+
+        if updated_active_tx_count > POLICY.maximum_thread_count {
+            return Err(TooManyActiveThreadCount);
+        }
+        if tx.addrs.is_empty() {
+            return Err(NoAddress);
+        }
+
+        let (is_congested, reset_counter) =
+            match updated_active_tx_count.cmp(&POLICY.congestion_threshold) {
+                Less if self.is_congested => (false, self.reset_counter + 1),
+                Equal | Greater if !self.is_congested => (true, self.reset_counter),
+                _ => (self.is_congested, self.reset_counter),
+            };
```

Comment:
Wondering the reason of determine congestion by number of txs instead of accumulated CUs?

Reply:
nice question. and really thanks for even taking look into the sim code. :)

accumulated cu could be a possibility. however, i think active number of running tx is more direct. and accumulated cu is hard to use to define congestion with very short feedback delay and could be skewed due to overpaid txes so far.

Reply:
this is definitely an open question. Appreciate for putting forward your thoughts.